### ISSUE TITLE
Discussion: Foundational / Pre-trained Models in Darts

### ISSUE THREAD

**Issue by daidahao** [unit8co/darts#2933]
**Is your feature request related to a current problem? Please describe.**
Amazon has just released [Chronos 2](https://github.com/amazon-science/chronos-forecasting), a pre-trained Transformer model for forecasting. Unlike previous Chronos, it has direct multivariate and (future and past) covariate support.

I wonder if there is any interest from the Darts side to implement and support it inside Darts? If so, I am happy to try implementing a prototype in Darts APIs.

**Describe proposed solution**
A prototype implementation of pre-trained Chronos 2 in Darts torch model APIs. Some additional dependencies may be required (from a glance of [Amazon code](https://github.com/amazon-science/chronos-forecasting/tree/main/src/chronos/chronos2)):
- einops (maybe replaced by torch operations).

The user would also need to download pre-trained model weights (120M parameters, around 450MB) from Amazon S3 bucket.

**Describe potential alternatives**
A clear and concise description of any alternative solutions or existing features that might solve it.

**Additional context**

**Comment by JuanCruzC97** [comment]
It would be great to have the darts interface available to evaluate this kind of pretrained models.

**Comment by dennisbader** [comment]
Hi @daidahao and @JuanCruzC97, and thanks for raising this issue. We are currently re-evaluating whether we could / should interface some of these models.

One or two years ago I was against adding them, but now that the models also support covariates and multivariate forecasting it looks different :)

Before implementing anything, I would first like to think about a couple of high level topics.
It would be very nice to get your feedback on this. We could also shape the plan for this together, if you like :)

- Where would we have this interface? I'm worried specifically about unit testing, where each time our CI/CD pipeline is run, it would have to re-download the models.
  - Would we keep it in the main `darts` repo?
  - Or would create a new repo such as `darts-fm`?
  - Or is there even a better solution which wouldn't burden the main repo too much?
- What models should we add? 
  - In my opinion, all models must be built on `pytorch`. Otherwise, we won't add them.
  - Chronos 2 looks good with rich covariates-, and multivariate forecasting support
  - TimesFM2.5, maybe? There is currently a PR open #2934, even though that one might be a bit too early for me (since we haven't yet made a clear plan)
  - ...?
- API wise, the models must inherit from `GlobalForecastingModel` (I wonder if we could wrap some of the pytorch-based models with our TorchForecastingModel)
  - I hope not much more is needed, except of the downloading capabilities
  - fit() might be used for fine-tuning the models (this would be great, but might be tricky to unify across different models. I would leave it as optional for now).

**Comment by daidahao** [comment]
Hi @dennisbader, thank you for your very prudent and well-thought out response. Yes, I am very glad to be shaping and contributing to this core feature with you and perhaps @JuanCruzC97 as well, :)

Over the past two months, our team at Oxford have ported and experimented with several open-sourced foundation models, in parallel to using Darts for our main forecasting task. Therefore, a lot of my suggestions here would be from my experience of interacting with those models.

> Where would we have this interface? I'm worried specifically about unit testing, where each time our CI/CD pipeline is run, it would have to re-download the models.

Most TSFMs if not all of them are larger than typical TFTs or RNNs, but usually smaller than 1B, at around 50M-500M. See [GIFT-EVAL](https://huggingface.co/spaces/Salesforce/GIFT-Eval) for a list of recent TSFMs. At that scale, some small-scale unit testing should not take too long on GitHub Actions runners, like forecasting with a small batch but not fine-tuning. I am in support of keeping TSFMs in the main repo (at least for now) to prevent issues like internal API changes, separate documentation, maintenance, etc.

However, testing time could stack up should we test all use cases of a TSFM (target only, target + past cov., target + future cov., etc.). In that case, I would like to propose a few ways to mitigate the testing time:

- Create a dummy tiny model for non-accuracy related unit tests, like in [Chronos 2 repo](https://github.com/amazon-science/chronos-forecasting/tree/main/test).
- Limit the number of "actual model" unit tests to 1 or 2 for validating the accuracy. Reduce the dataset size and batch size to reduce runtime per test.
- If the runtime is still very long, exclude TSFM unit tests from the workflows by default. Add workflow rules to only trigger unit tests when the TSFM code has changed or parent classes like `TorchForecastingModel ` have changed.

> What models should we add?

> In my opinion, all models must be built on pytorch. Otherwise, we won't add them.

I agree. Almost all TSFMs have PyTorch reference implementation, so that is fine (in 2025).

> Chronos 2 looks good with rich covariates-, and multivariate forecasting support
> TimesFM2.5, maybe?

I think Chronos 2 would be a good start as I've mentioned they have built-in multivariate and co-variate support. In terms of TimesFM 2.5, their covariate is on the [roadmap](https://github.com/google-research/timesfm) but not fully implemented yet. For other models, I would suggest prioritising models with multivariate and co-variate support since they are more likely to benefit Darts users. 

All in all, [GIFT-EVAL](https://huggingface.co/spaces/Salesforce/GIFT-Eval) and [fev-leaderboard](https://huggingface.co/spaces/autogluon/fev-leaderboard) list some of the most recent TSFMs. We can use them as a reference to keep track of progress in the field and to select the models.

> API wise, the models must inherit from GlobalForecastingModel (I wonder if we could wrap some of the pytorch-based models with our TorchForecastingModel)

Wrapping most models with `TorchForecastingModel` should be possible since they are fundamentally (no pun intended) torch models and would work with Lightning APIs. There are, however, key differences to other torch models

- `fit()` is not needed before calling `predict()`. We may need some workaround with the default logic of `ForecastingModel` where `fit()` must be called before `predict()`.
- A few extra dependencies are required for TSFMs. From my experience, most models would need `huggingface_hub` (not `transformers`) and `safetensors` for downloading and loading model weights. Some would require tensor library like `einops` but can be replaced with `einsum()` and `reshape()` in PyTorch. Thankfully, most of those dependencies are rather lightweight.
- Most TSFMs download weights from Hugging Face, but some like Chronos 2 from AWS S3 bucket.
- Many TSFMs have their own fixed quantile levels, and cannot be changed without fine-tuning.
- Many TSFMs use median sampling (#2912), or "quantile over quantiles" sampling rather than MC sampling. Validating the model outputs when auto-regression are involved is impossible due to different sampling methods.
- Many TSFMs have default "output chunk lengths" (horizon each step), which may be shorter than the user-defined. Performance when the horizon (with auto-regression) is longer is not guaranteed.

Given the differences and many model specifies, I wonder if you would prefer a unified `FoundationModel` that all TSFMs inherit from or having all TSFMs inheriting from `TorchForecastingModel`.

> fit() might be used for fine-tuning the models (this would be great, but might be tricky to unify across different models. I would leave it as optional for now).

Fine-tuning would be model-specific. Chronos 2 supports fine-tuning with quantile loss and seems compatible with `QuantileRegression` (at a glance of their codebase). We might not have such luck with other models if their loss function is different.

Looking forward to your further thoughts on this. Thanks!

**Comment by hrzn** [comment]
Just my 2 cents: I think it'd be great for Darts to host a few selected reference implementation of these models and make them available to our users. Looking at the ecosystem, Darts could play a helpful role of a "neutral" place where several of these (sometimes competing) models can live and be easily used under the common joint Darts API. => IMO that's a good idea.
When it comes to the specifics: would it be worth controlling / versioning also the model weights (maybe hosting them on a separate darts-controlled git repo, or bucket), just to mitigate the risk of external dependencies (e.g. if some externally controlled bucket is brought down), and to avoid putting what I assume might be multi-100MBs files on the main repo.

**Comment by nehalecky** [comment]
Hi @dennisbader, @daidahao and @hrzn! Apologies about delay, family has been hit with a bad 🦠 that knocked us offline and we're still recovering. Thank you all for opening up the discussion and engagement. Really great insights, perspectives and feedback. Give me a few days to incorporate all and will chime in soon with a response.  Happy to align on a design, interface and modify this implementation to get this across the line. 

In general, after a review of other time series libraries and lack of a unified API to allow wider accessibility to time series foundation models (TSFMs) application, I concluded Darts had a strong foundation, from an interface, implementation, and guidance for for aligning TSFMs across the board (in tune with @hrzn's note above)—exciting to see this goal delivered and more broad adoption of these very useful tools!

Final note: Chronos 2 was released right as this PR was shipped! So far, I'm impressed, and feel it likely sets a bar for the core capabilities at TSFM should have. I have a working implementation for integration it with Darts based off of this PR, and brushing it up now, but will let our efforts here finalize it!

**Comment by dennisbader** [comment]
Hi @daidahao, @hrzn, @nehalecky and thanks all for joining the discussion :) (sorry, I was a bit busy with two other projects over the last weeks).

> I think Chronos 2 would be a good start as I've mentioned they have built-in multivariate and co-variate support.

I agree with that.

> For other models, I would suggest prioritising models with multivariate and co-variate support since they are more likely to benefit Darts users.

This one I agree with as well. It seems like the new go-to is adding support for multivariate / covariates. I personally don't mind not adding each and every model that exists at the moment. But rather add more in the near future once they support more features.

> Most TSFMs download weights from Hugging Face, but some like Chronos 2 from AWS S3 bucket.

It looks to me like Chronos 2 is now also available on Hugging Face, see [here](https://huggingface.co/amazon/chronos-2).
@hrzn, I would say that for now, having the model weights available on HuggingFace is okay as it would be a quick win without a lot of maintenance from our side (instead of versioning them ourselves). If this will become a need in the future, we can also add it at a later stage. 

As a first version, it would be great to have a `FoundationModel` base which must support:

- it must be a global model (single & multiple series support)
- easy / straight forward way to (down)load a model either from hugging face / (other source?), or use the cached loaded model from disk.
- fitting: fit() must still be implemented and callable, even if we won't train the model. This is to provide a unified api with other models. Fit is also used to remember the dimensionality of the series for later sanity checks during inference. E.g.`model.fit(single_series).predict(n=12)` must work, as well as `model.fit(single_series).predict(n=12, series=another_series_with_same_dimensionality)`
- inference: predict() should support
  - uni-(/multi) variate forecasting:
  - probabilistic forecasting: direct quantiles and sampled (if model doesn't support sampled, we can use our tricks to sample from predicted quantiles)
    - the forecasts components/columns must follow our quantile naming conventions
 
With above, we should be able to support the entire Darts API (e.g. historical forecasts, backtest, ...)

What would be missing are the optimized routines for historical forecasts (e.g. what we get from `TorchForecastingModel`). In the end it would be great if we could somehow support it as well, but maybe it's not top prio for the beginning. 

Also, I've been following a bit the nice work of @AzulGarza on TimeCopilot which already supports many foundational models under one hood :) Maybe we could take some inspiration from there on how address certain aspects?

> Fine-tuning would be model-specific...

True, let's keep fine-tuning out of the current scope for now :)

**Comment by daidahao** [comment]
@dennisbader Thank you for the blueprint here. I will put forward a draft PR for Chronos 2 in the next few days.

**Comment by nehalecky** [comment]
### Foundation Models Integration - Implementation Ready for Review

Hi @dennisbader, @daidahao, @hrzn, and @JuanCruzC97! 

I'm relatively new to the Darts community, but I'm dedicated to seeing this foundation models integration through and would greatly welcome all guidance and feedback from the team here.

As I mentioned [earlier in the thread](https://github.com/unit8co/darts/issues/2933#issuecomment-2434186495), I've been working on a unified implementation based on the excellent design feedback from everyone. @daidahao, I saw you mentioned planning to put forward a draft PR as well - I'd love to see your approach and very much look forward to your feedback on this implementation!


### Links

**Design Document:** https://gist.github.com/nehalecky/4807ffe147321914387858dd4633c5c7  
** Pull Request:** https://github.com/unit8co/darts/pull/2942  
** Branch:** [`feature/foundation-models`](https://github.com/nehalecky/darts/tree/feature/foundation-models)

### Implementation Summary

This PR implements the unified architecture we've been discussing, providing Darts as a "neutral ground" for multiple TSFMs under a common API. It directly addresses all maintainer concerns raised throughout this thread.

**Models Included:**
- **TimesFM 2.5** (Google, 200M params): Zero-shot, 10 quantiles, univariate
- **Chronos 2** (Amazon, 120M params): Zero-shot, 21 quantiles, **multivariate + covariates + fine-tuning**

### Addresses All Concerns 

#### 1. CI/CD Overhead (@dennisbader)
> *"Models would download every time we run our CI pipelines"*

**Solution:** Lazy loading - models download on first `.model` property access, not at import. Tests use mocks. **Zero CI/CD impact.**

#### 2. fit() Semantics (@dennisbader)
> *"fit() must still be implemented and callable... Fit is also used to remember the dimensionality of the series for later sanity checks during inference."*

**Solution:** Dual-path fit() implementation:
- **Zero-shot path** (`lora_config=None`): Validates inputs, tracks dimensionality, enables unified API - no training
- **Fine-tuning path** (`lora_config` provided): Applies PEFT/LoRA and trains on provided data

Both `model.fit(series).predict(n=12)` and `model.fit(series).predict(n=12, series=another_series)` work as specified.

#### 3. Unified Base Class (@daidahao)
> *"I wonder if you would prefer a unified FoundationModel that all TSFMs inherit from"*

**Solution:** `FoundationForecastingModel` base class:
- Base provides: Lazy loading, device management, PEFT infrastructure, test patterns
- Subclasses implement: `_load_pretrained_model()` (~15 lines per model)
- Result: 27 lines of duplicate code eliminated

#### 4. Testing Strategy (@daidahao)
> *"Create a dummy tiny model for non-accuracy related unit tests"*

**Solution:** 48/48 tests passing
- Unit tests: Mock `from_pretrained()` (fast, no downloads)
- Integration tests: `@pytest.mark.slow` (run on-demand)

### Key Features

**Zero-shot prediction:**
```python
from darts.models import ChronosModel

chronos = ChronosModel(context_length=512)
forecast = chronos.predict(n=24, series=train)
```

**With covariates (Chronos 2):**
```python
forecast = chronos.predict(
    n=24,
    series=sales,
    past_covariates=weather_history,
    future_covariates=planned_promotions
)
```

**Quantile forecasting:**
```python
forecast = chronos.predict(
    n=24,
    series=train,
    quantiles=[0.1, 0.5, 0.9]  # Registry-based
)
```

**Fine-tuning with PEFT/LoRA (Chronos 2):**
```python
chronos = ChronosModel(lora_config={
    "r": 8,
    "lora_alpha": 16,
    "target_modules": ["q_proj", "v_proj"]
})

chronos.fit(
    series=training_data,
    past_covariates=covariates,
    prediction_length=24,
    num_steps=1000,
    learning_rate=1e-5
)

forecast = chronos.predict(n=24, series=test_series)
```

---

### Open Questions for Maintainers

1. **Repository structure:** Is `darts/models/forecasting/foundation/` acceptable?
2. **Dependencies:** Continue with lazy imports + optional extras (`darts[timesfm]`, `darts[chronos]`), or prefer unified `darts[foundation]`?
3. **Documentation:** Where should design/architecture docs live in the repo?

---

### Next Steps

I'm committed to iterating on this based on feedback from the community and dedicated to seeing this integration through to completion. The implementation is ready for review, and I'm happy to adjust based on:
- Architectural preferences
- API design decisions
- Testing strategies
- Documentation organization

@daidahao, I'm especially interested in your perspective given your experience porting TSFMs at Oxford - would love to hear your thoughts on the approach!

Looking forward to all your guidance and collaboration! 🚀

**Comment by JuanCruzC97** [comment]
Hi everyone!
Thanks for adding me to the discussion on integrating foundational models into Darts. I’ll take some time to get more practical experience with models like Chronos 2, but I’m glad to assist with reviewing and testing the implementations proposed here.

**Comment by nehalecky** [comment]
Hey everyone! A quick update on #2942. I've refactored to a simpler implementation, including:

- removed registry system, validation helpers, excess abstraction
- fixed Chronos 2 multivariate support
- restructured to 7 logical commits for easier review

I've also cleaned up the previously mentioned [design doc](https://gist.github.com/nehalecky/4807ffe147321914387858dd4633c5c7). Feedback is most welcome, and @daidahao if you're open to collaborating, I'd be grateful! 

cc @dennisbader

**Comment by daidahao** [comment]
Hi everyone, following the discussions here on foundation model, I've opened a PR for foundation model and Chronos-2 at #2944.

Feel free to play ground with the new APIs and give your feedback or comments for the PR. :)

For discussions on using Chronos-2 in Darts, I have opened a sub-issue at #2943.

@dennisbader @hrzn @JuanCruzC97 @nehalecky

**Comment by nehalecky** [comment]
@daidahao thanks and glad to see the foundation model pattern emerging in Darts in https://github.com/unit8co/darts/pull/2944! From all the conversation here, your native Chronos 2 implementation clearly aligns with a target design for Darts, which prefers to (good to know!) and I'll defer to that approach (and close out my PR https://github.com/unit8co/darts/pull/2942).

**On PyTorch 2.0:** +1 to upgrading from torch>=1.8.0. Beyond SDPA for Chronos-2, the entire ecosystem benefits from torch.compile() and 2+ years of optimizations. ✨ 

**On fine-tuning:** Besides the strong zero-shot performance (beautifully demonstrated in https://github.com/unit8co/darts/issues/2943), fine-tuning will be a really strong follow up feature, and another huge benefit TSFMs bring, as demonstrated in the Chronos 2 paper. If you and / or @dennisbader  are willing to give some initial guidance, I'd be up for taking a swing to  implementing such, after the dust settles with your PR. 

**IMO, an update to CONTRIBUTING.md** with a brief section on criteria to reimplement and maintain a model versus wrap existing packages would be beneficial, (e.g., StatsForecast is wrapped, but Chronos-2 is reimplemented). Suspect it's aligned with @daidahao [mention](https://github.com/unit8co/darts/pull/2942#issuecomment-3493639905) that the model's implementation is in PyTorch, which yes gives all the PyTorch Lightning niceties in Darts. 

That said, I strongly agree with @abdulfatir on [call out to expose repo_id as a public param](https://github.com/unit8co/darts/pull/2944#discussion_r2505869008), for the future proofing and smoother integration with HF Hub where model weights will live. I had something similar in the first design to with my PR (albeit naively), helping in goal of Darts providing a unified api for TSFMs!

cc @dennisbader 

and thanking abdulfatir publicly for actual feedback on the PR! 🙌

**Comment by daidahao** [comment]
Hi all, now that #2944 has been merged, Chronos-2 is implemented and supported in Darts unified APIs. For developers of the community, we also provide `FoundationModel` that serves as a base class for future PyTorch-native foundation models.

Thank you for all the discussions so far as they have helped shape the foundation model implementation. Feel free to continue sharing your views or feedback on this new feature here.

**Comment by daidahao** [comment]
Hi everyone, [`TimesFM2p5Model`](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.timesfm2p5_model.html#darts.models.forecasting.timesfm2p5_model.TimesFM2p5Model) has been provided as part of Darts 0.41.0 release. 🚀 Our `FoundationModel` class has stood the test and continues to serve as the base class for integrating foundation models.
